File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        boolean[] visited = new boolean[n];
        int answer = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                bfs(i, n, computers, visited);
                answer++; 
            }
        }
        return answer;
    }

    private void bfs(int start, int n, int[][] computers, boolean[] visited) {
        Queue<Integer> q = new ArrayDeque<>();
        visited[start] = true;      
        q.offer(start);

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int next = 0; next < n; next++) {
                
                if (computers[cur][next] == 1 && !visited[next]) {
                    visited[next] = true;   
                    q.offer(next);
                }
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.js
class Queue {
    constructor() {
        this.items = [];
    }
    enqueue(item) {
        this.items.push(item);
    }
    dequeue() {
        return this.items.shift();
    }
    
    isEmpty() {
        return this.items.length === 0;
    }
}

function bfs(start, n, computers, visited) {
    const q = new Queue();
    q.enqueue(start);
    while(!q.isEmpty()) {
        const cur = q.dequeue();
        for(let next = 0; next < n; next++) {
            if(computers[cur][next] === 1 && !visited[next]) {
                visited[next] = true;
                q.enqueue(next);
            }
        }
    }
}

function solution(n, computers) {
    const visited = Array(n).fill(false);
    let answer = 0;
    for(let i = 0; i < n; i++) {
        if(!visited[i]) {
            bfs(i, n, computers, visited);
            answer++;
        }
    }
    
    
    
    return answer;
}